src/linear_algebra/matrix/is_diag.lean

```diff
@@ -0,0 +1,185 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import linear_algebra.matrix.symmetric
+import linear_algebra.matrix.orthogonal
+import data.matrix.kronecker
+
+/-!
+# Diagonal matrices
+
+This file contains the definition and basic results about diagonal matrices.
+
+## Main results
+
+- `matrix.is_diag`: a proposition that stats a given square matrix `A` is diagonal
+                    (i.e. `∀ m n, i ≠ j → A m n = 0`).
+
+## Tags
+
+diag, diagonal, matrix
+-/
+
+namespace matrix
+
+variables {α β R n m : Type*}
+
+open function
+open_locale matrix kronecker
+
+/-- `A.is_diag` means square matrix `A` is a dianogal matrix: `∀ i j, i ≠ j → A i j = 0`. -/
+def is_diag [has_zero α] (A : matrix n n α) : Prop := ∀ ⦃i j⦄, i ≠ j → A i j = 0
+
+/-- Matrix `A` is diagonal iff there is a vector `d`
+    such that A is the diagonal matix generated by `d`. -/
+lemma is_diag_iff_eq_diagonal [has_zero α] [decidable_eq n] (A : matrix n n α) :
+  A.is_diag ↔ (∃ d, A = diagonal d) :=
+begin
+  split,
+  { intros h,
+    use (λ i, A i i),
+    ext i j,
+    by_cases g : i = j;
+    simp [@h i j, g] },
+  { rintros ⟨d, rfl⟩ i j h, simp [h] }
+end
+
+/-- Every unit matrix is diagonal. -/
+@[simp] lemma is_diag_unit [has_zero α] [subsingleton n] (A : matrix n n α) : A.is_diag :=
+by { intros i j h, have h' := subsingleton.elim i j, contradiction }
+
+/-- Every zero matrix is diagonal. -/
+@[simp] lemma is_diag_zero [has_zero α] : (0 : matrix n n α).is_diag :=
+λ i j h, rfl
+
+/-- Every identity matrix is diagonal. -/
+@[simp] lemma is_diag_one [decidable_eq n] [has_zero α] [has_one α] :
+  (1 : matrix n n α).is_diag :=
+λ i j, one_apply_ne
+
+@[simp] lemma is_diag.map [has_zero α] [has_zero β]
+{A : matrix n n α} (ha : A.is_diag) {f : α → β} (hf : f 0 = 0) :
+  (A.map f).is_diag :=
+by { intros i j h, simp [ha h, hf] }
+
+@[simp] lemma is_diag.neg [add_group α] {A : matrix n n α} (ha : A.is_diag) :
+  (-A).is_diag :=
+by {intros i j h, simp [ha h]}
+
+@[simp] lemma is_diag.add
```

Comment:
```suggestion
lemma is_diag.add
```
As a rule of thumb, lemmas that are not of the form `A \iff B` or `A = B` should not be tagged with `@[simp]`. There are some exceptions, but they are few.

Reply:
But in this case, if `simp` knows `A` is diag and so is `B`, it can close the goal for us automatically?

Reply:
Will the `simp` tag do anything bad?

Reply:
It will start trying to find a proof of `A.is_diag` and `B.is_diag`. In 99% of the cases, this will be a waste of time. So I think it shouldn't be in the global simp set. Users can add it when they need it.